<a href="https://colab.research.google.com/github/elyasncr/Curso_Machine_learning/blob/main/Selecao_de_Atributos.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

## Seleção de Características Importantes
*   Definir quais caracterísitcas, entre as "naturais" e as produzidas, são mais importantes para a performance do modelo.

### Maldição da Dimensionalidade
* A inclusão de muitas características do modelo deteriora sua performance, tornando o modelo super ajustado

******************************

#### Qual é o subconjunto de atributos que torna o modelo mais genérico?
* Atributos generalistas VS atributos específicos
* Atributos não correlacionados

#### Técnincas utilizadas para a seleção de atributos
* Força bruta --> forçar o dado a ser aquilo necessário
* Testes estatísticos: Anova, Chi-Square







In [1]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.naive_bayes import GaussianNB
from sklearn.metrics import accuracy_score
from sklearn.feature_selection import chi2, SelectKBest

In [2]:
#Chamando a nossa base de dados
anuncio = pd.read_csv('ad.data', header=None)

# Exibindo a quantidade de linhas e colunas
anuncio.shape

(3279, 1559)

In [3]:
# Alocando todas as colunas na variável X, exceto a última coluna
X = anuncio.iloc[:,:-1].values

# Alocando a última coluna na variável Y
y = anuncio.iloc[:,-1].values

In [4]:
# Valores alocados em X
X

array([[125.    , 125.    ,   1.    , ...,   0.    ,   0.    ,   0.    ],
       [ 57.    , 468.    ,   8.2105, ...,   0.    ,   0.    ,   0.    ],
       [ 33.    , 230.    ,   6.9696, ...,   0.    ,   0.    ,   0.    ],
       ...,
       [ 23.    , 120.    ,   5.2173, ...,   0.    ,   0.    ,   0.    ],
       [  0.    ,   0.    ,   0.    , ...,   0.    ,   0.    ,   0.    ],
       [ 40.    ,  40.    ,   1.    , ...,   0.    ,   0.    ,   0.    ]])

In [5]:
# Valores alocados em Y
y

array(['ad.', 'ad.', 'ad.', ..., 'nonad.', 'nonad.', 'nonad.'],
      dtype=object)

In [6]:
# Criando o treinamento e teste para X e Y
X_treinamento, X_teste, y_treinamento, y_teste = train_test_split(X,y, test_size=0.3, random_state=0)

In [8]:
# Usando o método de Gaus após o treinamento
modelo1 = GaussianNB()
modelo1.fit(X_treinamento, y_treinamento)

# Fazendo a previsão do modelo de Gaus e o teste de performance
previsoes1 = modelo1.predict(X_teste)
performance = accuracy_score(y_teste, previsoes1)

Atraves da performance estamos testando o que aconteceu com o modelo e a sua previsão.
Fazendo um comparativo entre o que aconteceu e o que estamos prevendo.

In [9]:
# Resultado da performance
print(performance)

0.7764227642276422


Com o resultado da performance obtivemos 77% de performance sobre o modelo de Gaus

In [11]:
# Selecionando atributos para um novo modelo de Gaus
selecao = SelectKBest(chi2, k=7)

# Retornando os atributos selecionados para a nova variável --> Passando X e Y para o Select selecionar os melhores dados p/ conjunto
X_novo = selecao.fit_transform(X,y)

In [14]:
# Novo conjunto de dados com os atributos separados
X_novo


array([[125.    , 125.    ,   1.    , ...,   1.    ,   0.    ,   1.    ],
       [ 57.    , 468.    ,   8.2105, ...,   1.    ,   0.    ,   0.    ],
       [ 33.    , 230.    ,   6.9696, ...,   0.    ,   0.    ,   0.    ],
       ...,
       [ 23.    , 120.    ,   5.2173, ...,   0.    ,   0.    ,   0.    ],
       [  0.    ,   0.    ,   0.    , ...,   0.    ,   0.    ,   0.    ],
       [ 40.    ,  40.    ,   1.    , ...,   0.    ,   0.    ,   0.    ]])

In [15]:
# Exibindo a quantidade de linhas e colunas
X_novo.shape

(3279, 7)

Se repararmos vemos que o conjunto de dados inicial tinha 1559 colunas, neste novo conjunto temos apenas 7.

In [16]:
# Metodo para exbir as colunas selecionas
print(selecao.get_support())

[ True  True  True ... False False False]


Como temos um grande número de colunas, e elas não têm nome, vemos somente True para coluna selecionada e False para não selecionada.

In [17]:
# Novo treinamento para o conjunto de dados X_novo
X_treinamento, X_teste, y_treinamento, y_teste = train_test_split(X_novo,y, test_size=0.3, random_state=0)

In [18]:
# Usando o método de Gaus após o treinamento do modelo novo
modelo2 = GaussianNB()
modelo2.fit(X_treinamento, y_treinamento)

# Fazendo a previsão do modelo de Gaus e o teste de performance  para o modelo novo
previsoes2 = modelo2.predict(X_teste)
performance2 = accuracy_score(y_teste, previsoes2)

In [19]:
# Resultado da performance
print(performance2)

0.9451219512195121


Com o resultado da performance do modelo 1 obtivemos 77% de performance sobre o modelo de Gaus.

Após selecionarmos os atributos para uma análise mais precisa, obtivemos um ganho de 17% em cima do modelo 1. Sendo assim, a performance do modelo 2 foi de 94%.